producer/src/oid4vc_core.rs

```diff
@@ -70,42 +67,47 @@ impl Verify for SecretManager {
                 DIDUrlQuery::from(&did_url),
                 Some(identity_iota::verification::MethodScope::VerificationMethod),
             )
-            .ok_or(Error::new(
-                ErrorKind::NotFound,
-                format!(
-                    "No verification method found for fragment=[{}]",
-                    did_url.fragment().unwrap()
-                ),
-            ))?;
-
-        let public_key_jwk = verification_method
-            .data()
-            .public_key_jwk()
-            .ok_or(Error::new(ErrorKind::NotFound, "No JWK found"))?;
-
-        let x = match public_key_jwk.params() {
-            identity_iota::verification::jwk::JwkParams::Okp(okp) => okp.x.as_str(),
-            identity_iota::verification::jwk::JwkParams::Ec(ec) => ec.x.as_str(),
-            identity_iota::verification::jwk::JwkParams::Rsa(_) => todo!(),
-            identity_iota::verification::jwk::JwkParams::Oct(_) => todo!(),
-        };
+            .ok_or(ProducerError::Generic(format!(
+                "No verification method found for fragment=[{}]",
+                did_url.fragment().unwrap()
+            )))?;
 
-        Ok(URL_SAFE_NO_PAD.decode(x.as_bytes()).unwrap())
+        // Try decode from `MethodData` directly, else use public JWK params.
+        verification_method
+            .data()
+            .try_decode()
+            .ok()
```

Comment:
Minor suggestion: this `ok()` converts the `Result` coming from `try_decode()` into an `Option` while later it is converted back to a `Result again here: `.ok_or(anyhow::anyhow!("Failed to decode public key for DID URL: {}", did_url))`.

I think this whole part can be rewritten like this:
```rust
        verification_method.data().try_decode().or_else(|_| {
            verification_method
                .data()
                .public_key_jwk()
                .and_then(|public_key_jwk| match public_key_jwk.params() {
                    JwkParams::Okp(okp_params) => Some(okp_params.x.as_bytes().to_vec()),
                    JwkParams::Ec(ec_params) => Some(ec_params.x.as_bytes().to_vec()),
                    _ => None,
                })
                .ok_or(anyhow::anyhow!("Failed to decode public key for DID URL: {}", did_url))
        })
```

Reply:
EDIT: including the base64 decoding it would/could look like this:
```rust
        verification_method.data().try_decode().or_else(|_| {
            verification_method
                .data()
                .public_key_jwk()
                .and_then(|public_key_jwk| match public_key_jwk.params() {
                    JwkParams::Okp(okp_params) => Some(okp_params.x.as_bytes().to_vec()),
                    JwkParams::Ec(ec_params) => Some(ec_params.x.as_bytes().to_vec()),
                    _ => None,
                })
                .ok_or(anyhow::anyhow!("Failed to decode public key for DID URL: {}", did_url))
                .and_then(|encoded_public_key| URL_SAFE_NO_PAD.decode(encoded_public_key).map_err(Into::into))
        })
```